pgscatalog_utils/scorefile/scoringfile.py

```diff
@@ -0,0 +1,201 @@
+                genome_build = header.HmPOS_build
+            else:
+                harmonised = False
+                genome_build = header.genome_build
+        else:
+            harmonised = False
+            genome_build = None
+
+        start_line, cols = get_columns(path)
+        is_wide = detect_wide(cols)
+
+        logger.info(f"Lazily reading variants from {path}")
+        variants = ScoringFile.read_variants(
+            path=path, start_line=start_line, fields=cols, name=name, is_wide=is_wide
+        )
+
+        # note: these generator expressions aren't doing a bunch of iterations
+        # it's just a data processing pipeline
+        variants = quality_control(
+            variants, header=header, harmonised=harmonised, wide=is_wide
+        )
+
+        return cls(
+            path=path,
+            header=header,
+            genome_build=genome_build,
+            harmonised=harmonised,
+            fields=cols,
+            variants=variants,
+            accession=name,
+        )
+
+    def generate_log(self, counted: typing.Counter):
+        log = {
+            key: str(value) if value is not None else None
+            for key, value in self.header.__dict__.items()
+        }
+
+        if log["variants_number"] is None:
+            # custom scoring files might not have this information
+            log["variants_number"] = counted["n_variants"]
+
+        if (
+            int(log["variants_number"]) != counted["n_variants"]
+            and not Config.drop_missing
+        ):
+            raise Exception(
+                f"Mismatch between variants_number and counted output {self.accession}"
+            )
+
+        # multiple terms may be separated with a pipe
+        if log["trait_mapped"]:
+            log["trait_mapped"] = log["trait_mapped"].split("|")
+
+        if log["trait_efo"]:
+            log["trait_efo"] = log["trait_efo"].split("|")
+
+        log["columns"] = self.fields
+        log["use_liftover"] = Config.liftover
+        log["use_harmonised"] = self.harmonised
+        log["sources"] = [k for k, v in counted.items() if k != "n_variants"]
+
+        return {self.accession: log}
+
+    @staticmethod
+    def read_variants(path, fields, start_line, name: str, is_wide: bool):
+        open_function = auto_open(path)
+        row_nr = 0
+
+        with open_function(path, mode="rt") as f:
+            for _ in range(start_line + 1):
+                # skip header
+                next(f)
+
+            while True:
+                batch = list(islice(f, Config.batch_size))
+                if not batch:
+                    break
+
+                csv_reader = csv.reader(batch, delimiter="\t")
+                yield from read_rows(csv_reader, fields, name, is_wide, row_nr)
+                # this is important for row_nr resets for each batch
+                row_nr += len(batch)
+
+
+def read_rows(csv_reader, fields: list[str], name: str, wide: bool, row_nr: int):
```

Comment:
Wondering if this will be useful/possible to have a dedicated class for the `variant`.
Could be more memory consuming thought, but it might help in term of code maintenance if the scoring/harmonized file formats change (or rename a column)

Reply:
Great idea 🤠 

I set up a simple class that [wraps a dictionary](https://github.com/PGScatalog/pgscatalog_utils/blob/stream_combine/pgscatalog_utils/scorefile/scorevariant.py) to keep things quick. 

I had to make the `__init__` very basic or it would bottleneck everything